test/integration_test/execute_rule/insufficient_tokenholder_funds.js

```diff
@@ -0,0 +1,149 @@
+  ExecuteRuleUtils = require('./utils'),
+  BN = require('bn.js'),
+  { Event } = require('../../test_lib/event_decoder');
+
+contract('TokenHolder::executeRule', async (accounts) => {
+
+  let accountProvider,
+    tokenHolder,
+    wallet1,
+    eip20TokenMock,
+    transferRule,
+    tokenRules,
+    ephemeralPrivateKey1,
+    ephemeralKeyAddress1,
+    keyData,
+    totalBalance = 200,
+    worker;
+
+  describe('ExecuteRule integration test', async () => {
+
+    it('Should fail when tokenholder has insufficient balance', async () => {
+
+      accountProvider = new AccountsProvider(accounts);
+
+      ( {
+        tokenHolder,
+        wallet1,
+        eip20TokenMock,
+        transferRule,
+        tokenRules,
+        worker,
+      } = await ExecuteRuleUtils.setup(accountProvider));
+
+      await eip20TokenMock.setBalance(tokenHolder.address, totalBalance);
+
+      ephemeralPrivateKey1 = '0xa8225c01ceeaf01d7bc7c1b1b929037bd4050967c5730c0b854263121b8399f3';
+      ephemeralKeyAddress1 = '0x62502C4DF73935D0D10054b0Fb8cC036534C6fb0';
+
+      let currentBlockNumber = await web3.eth.getBlockNumber(),
+        expirationHeight = currentBlockNumber + 50,
+        spendingLimit = 300;
+
+      await tokenHolder.submitAuthorizeSession(
+        ephemeralKeyAddress1,
+        spendingLimit,
+        expirationHeight,
+        { from: wallet1 },
+      );
+
+      keyData = await tokenHolder.ephemeralKeys(
+        ephemeralKeyAddress1,
+      );
+
+      let currentNonce = keyData.nonce,
+        amountTransferred = 250;
+
+      let nextAvailableNonce = currentNonce.toNumber() + 1;
+      const to = accountProvider.get();
+
+      // We are generating executable data to transfer 250 tokens.
+      // But, tokenholder has total 200 tokens.
+      const transferFromExecutable = await ExecuteRuleUtils.generateTransferFromExecutable(
+        tokenHolder.address,
+        to,
+        new BN(amountTransferred),
+      );
+
+      const { rsv } = await ExecuteRuleUtils.getExecuteRuleExTxData(
+        tokenHolder.address,
+        transferRule.address,
+        transferFromExecutable,
+        new BN(nextAvailableNonce),
+        ephemeralPrivateKey1,
+      );
+
+      const transactionResponse = await tokenHolder.executeRule(
+        transferRule.address,
+        transferFromExecutable,
+        (currentNonce.toNumber() + 1),
+        rsv.v,
+        EthUtils.bufferToHex(rsv.r),
+        EthUtils.bufferToHex(rsv.s),
+      );
+
+      const events = Event.decodeTransactionResponse(
+        transactionResponse,
+      );
+
+      // We should check against false here, however current version of web3
+      // returns null for false values in event log. After updating web3,
+      // this test might fail and we should use false (as intended).
+      assert.strictEqual(events[0].args['_status'], null);
+
+      assert.strictEqual(
+        (await eip20TokenMock.balanceOf(to)).cmp(new BN(0)),
+        0,
+      );
+
+      assert.strictEqual(
```

Comment:
This assert should be at the start since "It" statement says "TokenHolder has insufficient balance "

Reply:
It checks TH balance after executeRule was executed. It verifies TH balance to be same as it was before calling executeRule.